### PR TITLE
WIP: Dev/transform-next aka query2

### PR DESCRIPTION
**Pull request by johan-bjareholt**
Essentially a simple query language for querying buckets and transforming the data server-side which is much easier to read, can sort results and doesn't use the awkward chunking format anymore.

Also moved all query/transform related files to aw_transform instead of aw_core

TODO:
- [x] More transform unittests
- [x] aw-server endpoint
- [x] More query2 unittests
- [x] Chunking support for query2
- [x] More syntax error handling
- [x] datestart/dateend global variable for query_bucket
- [x] Implement API on aw-server
- [x] Remove debug prints
- [ ] Documentation on usage and available functions
- [x] ~~Typechecking on query2 functions~~ (will be added later when moving to aw-analysis)
- [x] Update WebUI
- [x] Write the TODO tests

### CODE REVIEW

File: aw_transform/query2.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+def parse(line, namespace):
+    separator_i = line.find("=")
+    var = line[:separator_i]
+    val = line[separator_i+1:]
+    if not val:
+        # TODO: Proper message
+        raise QueryException("Nothing to assign")
+    var, rest = _parse_token(var, namespace)
+    if var[0] is not Variable:
```
Comment (ErikBjare) [inline aw_transform/query2.py]: You probably mean `if not isinstance(var[0], Variable)`, right?

File: aw_transform/query2.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+        # TODO: Proper message
+        raise QueryException("Cannot assign to a non-variable")
+    val, rest = _parse_token(val, namespace)
+    # Parse token
+    var = var[0].parse(var[1], namespace)
+    val = val[0].parse(val[1], namespace)
+    return val, var
+
+def interpret(var, val, namespace, datastore):
+    if type(var) is Variable:
```
Comment (ErikBjare) [inline aw_transform/query2.py]: Similar to the typecheck before but this works. Still better to use `isinstance` though.

File: aw_transform/transforms.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+"""
+    Watcher specific transforms
+"""
+
+def split_url_events(events):
+    for event in events:
+        if "url" in event.data:
+            url = event.data["url"]
+            protocol_end = url.find('://')
```
Comment (ErikBjare) [inline aw_transform/transforms.py]: You should definitely be using [`urllib.parse`](https://docs.python.org/3.6/library/urllib.parse.html?highlight=urllib%20parse#module-urllib.parse) instead.

Comment (johan-bjareholt) [reply aw_transform/transforms.py]: fixed

File: aw_transform/transforms.py [hunk 4]
```diff
@@ -150,25 +150,19 @@ def limit_events(events, count):
...
-            path_end = path_start+url[path_start:].find('?')
-            if path_end < path_start:
-                path_end = len(url)
-            #print("Path: {}->{}".format(path_start, path_end))
-            event.data["protocol"] = url[:protocol_end]
-            event.data["domain"] = url[domain_start:domain_end]
-            event.data["path"] = url[domain_end:path_end]
-            event.data["options"] = url[path_end+1:]
+            parsed_url = urlparse(url)
+            event.data["protocol"] = parsed_url.scheme
```
Comment (ErikBjare) [inline aw_transform/transforms.py]: I'd like to suggest the following instead:

```
urlp = urlparse(url)
event.data.update({
    "protocol": urlp.scheme,
    "netloc": urlp.netloc if (not urlp.netloc[:4] == "www.") else urlp.netloc[4:],
    "path": urlp.path,
    "query": urlp.query,
    "fragment": urlp.fragment, 
})
```

As you see I also changed "options" -> ["query"](https://en.wikipedia.org/wiki/Query_string), "identifier" -> ["fragment"](https://en.wikipedia.org/wiki/Fragment_identifier), and "domain" -> "netloc" as these are better names that more accurately represent what they contain. It also has the added benefit of being almost identical to the urlparse object. (an argument for also changing "protocol" -> "scheme")

**Comment by ErikBjare** [comment]
Amazing work! :star:

**Comment by ErikBjare** [comment]
The same thing that I wrote [here](https://github.com/ActivityWatch/aw-server/pull/35#issuecomment-340752237) applies for this PR. This time its ok, but might want to keep it in mind for the next time a similar change is made.

**Comment by johan-bjareholt** [comment]
So the idea itself seems to be a good idea and is ready for production, have used it for a few weeks and works fine. The API however is questionable since it is not very REST-compatible.

These are our options:

### A1: Raw string (current solution)

Just a raw string query with metadata in the query

Example: /api/0/query/

```
STARTTIME=2017
ENDTIME=2018
CHUNK=FALSE
NAME="queryname"
events=query_bucket("bucketname")
events=some_filter(events, arg)
RETURN=events
```
NAME could also possibly be moved to URL

### A2: Raw string with metadata in URL

Example: /api/0/query/queryname/starttime/endtime/?chunk=(1|0)

```
events=query_bucket("bucketname")
events=some_filter(events, arg)
RETURN=events
```

### B1: JSON array for queries with metadata in JSON aswell

Example: /api/0/query/queryname/

```
{
    "NAME": "queryname",
    "starttime": "2017",
    "endtime": "2018",
    "chunk": false,
    "query": [
      "events=query_bucket(\"bucket_name\")",
      "events=some_filter(events, arg)",
      "return=events",
    ]
}
```

### B2: JSON array for queries with metadata in URL

Example: /api/0/query/queryname/starttime/endtime/?chunk=(1|0)


```
[
  "events=query_bucket(\"bucket_name\")",
  "events=some_filter(events, arg)",
  "return=events",
]
```

### C 1&2: Multiline JSON string with metadata in URL or in JSON

Ugly escaping

Example1: /api/0/query/
Example2: /api/0/query/queryname/starttime/endtime/?chunk=(1|0)

```
{
    /* Only if C1
    "NAME": "queryname",
    "starttime": "2017",
    "endtime": "2018",
    "chunk": false,
    */
    "query": "\n\
      events=query_bucket(\"bucket_name\") \n\
      events=some_filter(events, arg) \n\
      return=events \
    "
}
```

### D 1&2: Make both A and B available

I personally think this is a bad idea

# Summary

This is a Request for Comments. I none of the alternatives are perfect, so we simply just have to choose one.

The A alternative is bad because it's not JSON so it can't be marshalled by flask-restplus, and multiline strings are only supported in newer javascript versions so it looks quite ugly (can be solved by using babel).
B is bad because it's quite annoying to write each query on a row each as well as having to do escaping on quotation marks.
C is bad because it takes the bad parts of both A and B into a single thing.
D is bad because it will be annoying to maintain and it should be preferably only be one API to do one thing.

**Comment by ErikBjare** [comment]
I actually prefer B2, with some modification:

Example: /api/0/query?name={name}&start={start}&end={end}&cache=(1|0)

```json
{
  "query": [
    "events=query_bucket('bucket_name')",
    "events=some_filter(events, arg)",
    "return=events",
  ]
}
```

For the query itself I wouldn't mind this:

```json
{
  "query": "events=query_bucket('bucket_name'); \
            events=some_filter(events, arg); \
            return=events;"
}
```

Also, a bit unrelated, but would this be a valid query?

```python
events = query_bucket('bucket_name');
return = events;
```

Differences: Single quotes instead of double quotes around `'bucket_name'`, spaces around `=` operator, semicolons (instead of)/(in addition to) linebreaks.

**Comment by johan-bjareholt** [comment]
Now with even more coverage!

**Pull request merged**